models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py

```diff
@@ -0,0 +1,809 @@
+            data_indices=data_indices,
+            statistics=statistics,
+            graph_data=graph_data,
+            truncation_data=truncation_data,
+        )
+
+        self.get_noise_schedule = diffusion_samplers.get_noise_schedule
+
+        self.noise_embedder = SinusoidalEmbeddings(num_channels=self.noise_channels, max_period=1000)
+
+        self.noise_cond_mlp = nn.Sequential()
+        self.noise_cond_mlp.add_module("linear1_no_gradscaling", nn.Linear(self.noise_channels, self.noise_channels))
+        self.noise_cond_mlp.add_module("activation", nn.SiLU())
+        self.noise_cond_mlp.add_module("linear2_no_gradscaling", nn.Linear(self.noise_channels, self.noise_cond_dim))
+
+    def _calculate_input_dim(self, model_config):
+        base_input_dim = super()._calculate_input_dim(model_config)
+        return base_input_dim + self.num_output_channels  # input + noised targets
+
+    def _assemble_input(self, x, y_noised, bse, grid_shard_shapes=None, model_comm_group=None):
+        node_attributes_data = self.node_attributes(self._graph_name_data, batch_size=bse)
+        if grid_shard_shapes is not None:
+            shard_shapes_nodes = self._get_shard_shapes(node_attributes_data, 0, grid_shard_shapes, model_comm_group)
+            node_attributes_data = shard_tensor(node_attributes_data, 0, shard_shapes_nodes, model_comm_group)
+
+        # combine noised target, input state, noise conditioning and add data positional info (lat/lon)
+        x_data_latent = torch.cat(
+            (
+                einops.rearrange(x, "batch time ensemble grid vars -> (batch ensemble grid) (time vars)"),
+                einops.rearrange(y_noised, "batch ensemble grid vars -> (batch ensemble grid) vars"),
+                node_attributes_data,
+            ),
+            dim=-1,  # feature dimension
+        )
+        shard_shapes_data = self._get_shard_shapes(x_data_latent, 0, grid_shard_shapes, model_comm_group)
+
+        return x_data_latent, None, shard_shapes_data
+
+    def _assemble_output(self, x_out, x_skip, batch_size, ensemble_size, dtype):
+        bse = batch_size * ensemble_size
+        x_out = einops.rearrange(x_out, "(bse n) f -> bse n f", bse=bse)
+        x_out = einops.rearrange(x_out, "(bs e) n f -> bs e n f", bs=batch_size).to(dtype=dtype)
+
+        for bounding in self.boundings:
+            # bounding performed in the order specified in the config file
+            x_out = bounding(x_out)
+        return x_out
+
+    def _make_noise_emb(self, noise_emb: torch.Tensor, repeat: int) -> torch.Tensor:
+        out = einops.repeat(
+            noise_emb, "batch ensemble noise_level vars -> batch ensemble (repeat noise_level) vars", repeat=repeat
+        )
+        out = einops.rearrange(out, "batch ensemble grid vars -> (batch ensemble grid) vars")
+        return out
+
+    def _generate_noise_conditioning(self, sigma: torch.Tensor, edge_conditioning: bool = False) -> torch.Tensor:
```

Comment:
How easy do you think would be to adapt this in the future if we have say many encoder or a different graph than one with data-hidden-data?